Configurability using config file

Victor-Schwan opened the pull request:
BEGINRELEASENOTES
- Paths to personal and detector-specific files/folders outsourced into the config file
- Choose config file with `argparse`
- Path operations using `pathlib`
- so far, only tested with `condorJobs_sim.py`

ENDRELEASENOTES

Victor-Schwan committed:
variables outsourced to config.py; condorJobs_sim runs until actual execution at the end

Victor-Schwan committed:
switch from str paths to Path objects

Victor-Schwan committed:
minor fix

Victor-Schwan committed:
reco: switched to pathlib, more config vars added; conffig: environment moved here

Victor-Schwan committed:
time stuff removed; wait slowed down the execution significantly, maybe perf_counter_ns() is better without wait

Victor-Schwan committed:
job number calc in verbose mode fixed

Victor-Schwan committed:
verbose mode for debugging of slow loop removed

Victor-Schwan committed:
minor refactoring Condor/condorJobs_sim.py

Victor-Schwan committed:
minor

Victor-Schwan committed:
added afs paths to config

Victor-Schwan committed:
mostly chhanges to paths

Victor-Schwan committed:
condorJobs_sim: argparse to choose config file,
introduced functions;
cld_config added for cld specific variables

Victor-Schwan committed:
condorJobs_sim: improved argparsing: .py suffix can be omitted

Zehvogel commented on TrackingPerformance/Condor/condorJobs_sim.py:
```diff
@@ -1,156 +1,248 @@
+    return parser.parse_args()
+
+
+def load_config(config_name):
+
+    config_file = Path(config_name)
+
+    # Ensure the config file has the .py extension
+    if config_file.suffix != ".py":
+        config_file = config_file.with_suffix(".py")
+
+    # Create an absolute path
+    config_path = config_file.resolve()
+
+    # Ensure the file exists
+    if not config_path.exists():
+        raise FileNotFoundError(f"The configuration file {config_path} does not exist.")
+
+    spec = importlib.util.spec_from_file_location("config", config_path)
+    config = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(config)
+    return config
```
this looks familiar, @tmadlener do we have this already importable from somewhere in key4hep and if not should we make that happen?

tmadlener replied:
We were discussing putting [this bit](https://github.com/iLCSoft/ILDConfig/blob/master/StandardConfig/production/py_utils.py#L9-L49) into the k4FWCore utils, but ILDReconstruction was the only use case at the time. We could move it to k4FWCore.utils

Victor-Schwan replied:
It is not entirely the same. We could tackle this in a later refactoring

Zehvogel replied:
Fine with me. I just wanted to avoid having to many copies of the same thing flying around. :)

Zehvogel commented on TrackingPerformance/Condor/condorJobs_sim.py:
```diff
@@ -1,156 +1,248 @@
+    # It will check if the ouputs exist and contain correct number of events
+    # if not it will send job to rerun simulation
+
+    JOB_FLAVOR = "testmatch"
+    # Job flavours:
+    #   espresso     = 20 minutes
+    #   microcentury = 1 hour
+    #   longlunch    = 2 hours
+    #   workday      = 8 hours
+    #   tomorrow     = 1 day
+    #   testmatch    = 3 days
+    #   nextweek     = 1 week
```
I know it was already `testmatch` before but does it really take that long? I think we might want to aim for job sizes closer to `longlunch`

We also want to be able to set this from the config.

Victor-Schwan replied:
Fair point. Done

Zehvogel commented on TrackingPerformance/Condor/condorJobs_sim.py:
```diff
@@ -1,156 +1,248 @@
+            output_dir = SIMEOSDir / part
+            output_dir.mkdir(parents=True, exist_ok=True)
+            output_file_path = output_dir / output_file_name
+            if CHECK_OUTPUT and output_file_path.exists():
+                root_file = ROOT.TFile(output_file_path, "READ")
+                events_tree = root_file.Get("events")
+                if events_tree:  # FIXME: why no else?
+                    if events_tree.GetEntries() == int(Nevt_per_job):
+                        root_file.Close()
+                        continue
+                root_file.Close()
+            else:
+                need_to_create_scripts = True
```
This check looks like we only rewrite the files when we change the number of events per job? So if we just change the type of particles they are not written?

nitasad replied:
From what I can see, there doesn't seem to be much change compared to what I did. This check will rerun the simulation for the missing files because the jobs crashed or encountered other issues, which happens quite often. Therefore, you need to use the same variables. It will also verify if the output files contain the correct number of events. I've encountered this issue several times where the files exist but contain only a few events.

Victor-Schwan replied:
To me, it seems like the `need_to_create_scripts` variable is only set to `True` if the condition in line `161` equals `False`. The if statement `events_tree.GetEntries() == int(Nevt_per_job)` misses an `else` to ensure that missing events are produced. I will create an issue based on it to keep this pull request cleaner

Zehvogel commented on TrackingPerformance/Condor/condorJobs_sim.py:
```diff
@@ -1,156 +1,248 @@
+                f" --compactFile $k4geo_DIR/{config.detModPaths['ILD_l5_v11']}",  # Note the change to use double quotes for the dictionary key
+                f"--outputFile {output_file_name}",
+                f"--steeringFile {config.sim_steering_file}",  # "CLDConfig/CLDConfig/cld_steer.py "
+                # TODO: FIX
+                # "--enableGun",
+                # f"--gun.particle {part}-",
+                # f"--gun.energy {momentum}*GeV",
+                # "--gun.distribution uniform",
+                # f"--gun.thetaMin {theta}*deg",
+                # f"--gun.thetaMax {theta}*deg",
+                # "--crossingAngleBoost 0",
+                # f"--numberOfEvents {Nevt_per_job}",
```
Does something break when commenting it in? Without it the file is very useless :)

Victor-Schwan replied:
done

Zehvogel commented on TrackingPerformance/Condor/condorJobs_sim.py:
```diff
@@ -1,156 +1,248 @@
+            bash_script = (
+                "#!/bin/bash \n"
+                f"source {environ_path} \n"
+                "git clone https://github.com/Victor-Schwan/TrackingStudies.git \n"  # FIXME: new repo location to config (to switch easily to local fork)
```
Is this needed (or even correct)? Before we were cloning `CLDConfig` here to have the `cld_steer.py` file available.

Victor-Schwan replied:
I only changed the Path. Maybe @gaswk can comment on this?

Zehvogel replied:
You would have "only changed the path" if this was `github.com/gaswk/FullSim` but it was `github.com/gaswk/CLDConfig`

nitasad replied:
`CLDConfig` is required to use `cld_steer.py`. I use my version (`https://github.com/gaswk/CLDConfig.git`) because it includes configurable values for vertex resolution. However, it is preferable to use the official one `https://github.com/key4hep/CLDConfig`.

Victor-Schwan replied:
If I understand correctly, this is superfluous. @gaswk you presumably cloned your fork from `CLDConfig` to `afs`, right? Then the job should be able to access it without having to clone it again if I am not mistaken:
"This has access to two shared file systems, AFS and EOS. The batch system can also see both filesystems."([batchdocs](https://batchdocs.web.cern.ch/concepts/dataflow.html))

nitasad replied:
Yes, it has access to `afs` and can use your own fork from there. However, I think it is preferable to use the steering file from `https://github.com/key4hep/` by default. We could also make it configurable, allowing you to specify a different steering file if desired?

Victor-Schwan replied:
In `cld_config` line 26, one can specify the steering file:
```
sim_steering_file = detectorDIR / "CHANGE" / "PATH"  # FIXME
```
The one from `https://github.com/key4hep/` should be accessible via `cvmfs`, shouldn't it?

Zehvogel replied:
> The one from `https://github.com/key4hep/` should be accessible via `cvmfs`, shouldn't it?

It should be accessible in the future but not at the moment. The ILD equivalent [`ddsim_steer.py`](https://github.com/iLCSoft/ILDConfig/blob/master/StandardConfig/production/ddsim_steer.py) is also not available (yet) on cvmfs as far as I know...

tmadlener replied:
Technically the tags are available under `/cvmfs/ilc.desy.de/sw/ILDConfig/vxx-yy-zz/`, but not within a key4hep release yet.

Victor-Schwan replied:
I removed the cloning. If you want another solution, I suggest to open a new issue

Zehvogel commented on TrackingPerformance/Condor/condorJobs_sim.py:
```diff
@@ -1,156 +1,248 @@
+    # ====================
+    system(
+        "cd " + str(directory_jobs) + "; condor_submit condor_script.sub"
+    )  # FIXME: use subprocess instead?
```
good idea :)
or maybe even get rid of this or make it optional so that one can inspect the file before submission.
I think we can expect a user to be able to still type this one line by hand. If we want to be extra nice we could even print a small message from which this is directly copy-pasteable.

Victor-Schwan replied:
Or we could add a `no-submit` flag to `argparse`?

Zehvogel replied:
yes a flag for this would also be a good idea but I am unsure if the default should be to submit it. :shrug:

Victor-Schwan replied:
Another option is to outsource the execution and move it to a new script. This script can call the `main` function of `condorJobs_sim` and then submit the jobs.

Victor-Schwan replied:
Moved to separate issue

Zehvogel reviewed (commented):
Thx! 

Some comments below. I did not look much at the other files yet.

Victor-Schwan committed:
parse_arg and load_config outsourced to utils; easily reusable in e.g. condorJobs_reco

Victor-Schwan committed:
JOB_FLAVOR outsourced to config

tmadlener commented on TrackingPerformance/Condor/condorJobs_reco.py:
```diff
@@ -109,29 +145,37 @@
             "#!/bin/bash \n"
-            f"source {setup} \n"
-            "git clone https://github.com/gaswk/CLDConfig.git \n"
+            f"source {environ_path} \n"
+            "git clone https://github.com/gaswk/CLDConfig.git \n"  # FIXME: new repo location to config (to switch easily to local fork)
```
Do you want to keep the cloning in the batch jobs? Talking mostly about ILDConfig now, but these can become quite large and cloning can take quite a bit. If it's at all possible to lift the cloning to a "general setup step" that happens before the submission, etc. happens, I would recommend doing that. The obvious disadvantage is that it becomes user responsibility to make sure that things are consistent and that the config that was used for submission is reproducible.

Potentially something for another PR.

Victor-Schwan committed:
config gun args actually incorporated in ddsim command;
clone in `bash_script` in `condorJobs_sim` removed

Victor-Schwan committed:
reco: adde cfile from

Victor-Schwan committed:
renamed config template

Victor-Schwan committed:
minor addendum config template

Zehvogel commented:
@Victor-Schwan is this ready now?

Victor-Schwan commented:
Not ready yet. When I ran it on `lxbatch` there was a permission problem in `sim_jobs` which needs to be debugged. And I have an almost completed commit waiting removing double definitions of vars and some other refacs

Zehvogel commented:
Ok

Victor-Schwan committed:
double definitions of vars removed:
now only definied in config file and not anymore again in scripts;
integer variables defined as `ints` and not `strings`:
unnecessary type casting removed

Victor-Schwan committed:
number of events specified in config is ceiled instead of floored (rounding when dividing Nevts to jobs)

Victor-Schwan committed:
Alignment of names in  and

Victor-Schwan committed:
refac, alignment of naming and structure in sim and reco

Victor-Schwan committed:
fix path for xrootd

Victor-Schwan committed:
minor fix; refacs: mainly renaming vars

Zehvogel commented on TrackingPerformance/Condor/condorJobs_reco.py:
```diff
@@ -1,167 +1,223 @@
+            input_file = (
+                sim_eos_dir / part / input_file_path
+            )  # FIXME: reasonable that part is twice in the path?
+
```
I think having a directory per particle type is useful :)

Zehvogel commented on TrackingPerformance/Condor/condorJobs_reco.py:
```diff
@@ -1,167 +1,223 @@
+    # It will check if the ouputs exist and contain correct number of events
+    # if not it will send job to rerun reconstruction
+
+    # Set default value if ResVDX_UV_ is not defined or empty
+    try:
+        if not ResVDX_UV_:
+            ResVDX_UV_ = ["0.003"]
+    except NameError:
+        ResVDX_UV_ = ["0.003"]
```
I know this is from Gaelle but on the one hand I am completely confused what is supposed to hapen here syntax wise, while on the other hand we should make the whole `ResVDX_UV_` optional (also below).

It would be nice to be able to specify additional arguments to k4run in the config and then this could go there.

Victor-Schwan replied:
Moved to a new issue

Zehvogel commented on TrackingPerformance/Condor/condorJobs_sim.py:
```diff
@@ -1,156 +1,212 @@
+            # Build ddsim command
+            arguments = [
+                # f" --compactFile /afs/cern.ch/work/g/gasadows/k4geo/FCCee/CLD/compact/{DetectorModelList_[0]}_3T/{DetectorModelList_[0]}.xml "
+                f" --compactFile $k4geo_DIR/{config.det_mod_paths['ILD_l5_v11']}",  # Note the change to use double quotes for the dictionary key
```
:warning:

Victor-Schwan replied:
`det_mod_paths` key is in the next commit chosen from `config.detector_model_list[0]`

Zehvogel reviewed (commented):
Some more comments.

I guess we will need a future PR on `simJobs_reco` to make it a bit more flexible but for now I am happy.

I will merge after @gaswk approves. :)

Victor-Schwan committed:
sim: output file rm on afs; minor fixes; make Plotting/analysis_tracking compatible with general changes

nitasad commented on TrackingPerformance/Condor/condorJobs_reco.py:
```diff
@@ -1,167 +1,223 @@
+            # Check if the output file already exists and has correct Nb of events
+            output_dir = rec_eos_dir / part
+            output_dir.mkdir(parents=True, exist_ok=True)
+            output_file = (output_dir / output_file_name).with_suffix("_edm4hep.root")
```
Should be `.edm4hep.root` to be consistent

nitasad commented on TrackingPerformance/Condor/condorJobs_reco.py:
```diff
@@ -1,167 +1,223 @@
+
+            arguments = (
+                f" --GeoSvc.detectors=$K4GEO/FCCee/CLD/compact/{config.detector_model_list[0]}/{config.detector_model_list[0]}.xml"
+                + " --inputFiles "
+                + input_file
+                + " --outputBasename  "
+                + output_file_name
+                + f" --VXDDigitiserResUV={ResVDX_UV_[0]}"
+                + " --trackingOnly"
+                + " -n "
+                + config.N_EVTS_PER_JOB
+            )
```
`input_file` and `config.N_EVTS_PER_JOB` need to be of type str

nitasad commented on TrackingPerformance/Condor/condorJobs_reco.py:
```diff
@@ -1,167 +1,223 @@
+            bash_script = (
+                "#!/bin/bash \n"
+                f"source {config.setup} \n"
+                "git clone https://github.com/gaswk/CLDConfig.git \n"  # FIXME: see issues
```
Please use https://github.com/key4hep/CLDConfig.git and I think `ResVDX_UV_` argument should be removed, at least for now

Victor-Schwan committed:
changes to reco requested by Gaelle

Victor-Schwan committed:
ResVDX_UV_ removed from reco as requested by Gaelle

Victor-Schwan committed:
Revert Plotting/analysis_tracking.py to state before Victor's PR

Victor-Schwan committed:
reco: change aida file suffix to .aida.root

Zehvogel commented:
@gaswk is everything fine from your side now?

gaswk commented:
> @gaswk is everything fine from your side now?

I am waiting for your response to my previous comment regarding the naming conventions for the output files. To maintain consistency, should we use `_` instead of `.` for all output names? Currently, the outputs from the reco are named `_edm4hep.root` and `_aida.root` by default.

Zehvogel commented:
> I am waiting for your response to my previous comment regarding the naming conventions for the output files. To maintain consistency, should we use `_` instead of `.` for all output names? Currently, the outputs from the reco are named `_edm4hep.root` and `_aida.root` by default.

Oops, sorry for missing that. I would prefer `.` instead of `_` but if `CLDConfig` still produces `_` we would need to change it there first of course :/ I think a have a change for that ready somewhere but that also changes some more parts.

@Victor-Schwan opposite to previous instructions I would then suggest that you change back to `_` for this PR and we make the change to `.` later...

Victor-Schwan committed:
minor: typehint main functions

Victor-Schwan committed:
config option EDM4HEP_SUFFIX_WITH_UNDERSCORE added and path building in condorJobs_sim and _reco adapted

Victor-Schwan commented:
I added a bool var in the config file which switches between `_edm4hep.root` and `.edm4hep.root` paths. What do you think?

Zehvogel commented on TrackingPerformance/config_template.py:
```diff
@@ -0,0 +1,63 @@
+nightlies = Path("/cvmfs/sw-nightlies.hsf.org/key4hep/setup.sh")
+setup = nightlies  # choose either stable or nightlies
+
+# False: filepath_edm4hep.root; False: filepath.edm4hep.root
+EDM4HEP_SUFFIX_WITH_UNDERSCORE = False
```
```suggestion
# True: filepath_edm4hep.root; False: filepath.edm4hep.root
EDM4HEP_SUFFIX_WITH_UNDERSCORE = False
```

Victor-Schwan replied:
yes, thanks

Zehvogel commented on TrackingPerformance/config_v.py:
```diff
@@ -0,0 +1,63 @@
+nightlies = Path("/cvmfs/sw-nightlies.hsf.org/key4hep/setup.sh")
+setup = nightlies  # choose either stable or nightlies
+
+# False: filepath_edm4hep.root; False: filepath.edm4hep.root
```
```suggestion
# True: filepath_edm4hep.root; False: filepath.edm4hep.root
```

Zehvogel reviewed (changes requested):
I think you meant this?

Victor-Schwan committed:
minor fixup

gaswk commented:
Everything is fine for me now!
